File: src/main/rooms/Game.java
package main.rooms;

import main.Language;
import main.buttons.*;
import main.buttons.iconbuttons.Exit;
import main.buttons.iconbuttons.IconButton;
import main.buttons.iconbuttons.Save;
import main.combos.Combo;
import org.apache.commons.collections4.CollectionUtils;
import processing.core.PConstants;
import processing.data.JSONArray;
import processing.data.JSONObject;

import javax.swing.*;
import java.time.Duration;
import java.time.LocalDateTime;
import java.util.*;

public class Game extends Room {

    private final TreeMap<Group, ArrayList<Element>> discovered = new TreeMap<>(Group::compareTo);
    private final ArrayList<Combo> history = new ArrayList<>();

    private SaveFile saveFile;

    private final Pane success;
    private final Pane hint;
    private final Save save;
    private final Exit exit;
    private final IconButton historyButton;
    private final IconButton hintButton;
    private final IconButton undoButton;

    private final Arrow groupLeftArrow;
    private final Arrow groupRightArrow;
    private final Arrow elementAUpArrow;
    private final Arrow elementADownArrow;
    private final Arrow elementBUpArrow;
    private final Arrow elementBDownArrow;

    private LocalDateTime hintTime; //time of next hint

    private boolean gameLoaded = false;
    private Element hintElement;

    public String mode;

    public Game() {
        this.success = new Pane() {
            @Override
            protected String getText() {
                return Language.getLanguageSelected().getLocalizedString("game", "you created");
            }
        };
        this.hint = new Pane() {
            @Override
            protected String getText() {
                return Language.getLanguageSelected().getLocalizedString("game", "element hint");
            }
        };
        this.save = new Save();
        this.exit = new Exit();

        this.groupLeftArrow = new Arrow(Arrow.LEFT) {
            @Override
            protected boolean canDraw() {
                return Group.pageNumber > 0;
            }

            @Override
            public void clicked() {
                if (this.canDraw()) {
                    Group.pageNumber--;
                }
            }
        };

        this.groupRightArrow = new Arrow(Arrow.RIGHT) {
            @Override
            protected boolean canDraw() {
                return Group.pageNumber < Group.totalPages - 1;
            }

            @Override
            public void clicked() {
                if (this.canDraw()) {
                    Group.pageNumber++;
                }
            }
        };

        this.elementAUpArrow = new Arrow(Arrow.UP) {
            @Override
            protected boolean canDraw() {
                return Group.groupSelectedA != null && Element.pageNumberA > 0;
            }

            @Override
            public void clicked() {
                if (this.canDraw()) {
                    Element.pageNumberA--;
                }
            }
        };

        this.elementADownArrow = new Arrow(Arrow.DOWN) {
            @Override
            protected boolean canDraw() {
                return Group.groupSelectedA != null && Element.pageNumberA < Element.totalPagesA - 1;
            }

            @Override
            public void clicked() {
                if (this.canDraw()) {
                    Element.pageNumberA++;
                }
            }
        };

        this.elementBUpArrow = new Arrow(Arrow.UP) {
            @Override
            protected boolean canDraw() {
                return Group.groupSelectedB != null && Element.pageNumberB > 0;
            }

            @Override
            public void clicked() {
                if (this.canDraw()) {
                    Element.pageNumberB--;
                }
            }
        };

        this.elementBDownArrow = new Arrow(Arrow.DOWN) {
            @Override
            protected boolean canDraw() {
                return Group.groupSelectedB != null && Element.pageNumberB < Element.totalPagesB - 1;
            }

            @Override
            public void clicked() {
                if (this.canDraw()) {
                    Element.pageNumberB++;
                }
            }
        };

        this.historyButton = new IconButton("resources/images/history_button.png") {
            @Override
            public void clicked() {
                main.switchRoom(main.historyRoom);
            }
        };

        this.hintButton = new IconButton("resources/images/hint_button.png") {
            @Override
            public void clicked() {
                main.switchRoom(main.hintRoom);
            }
        };

        this.undoButton = new IconButton("resources/images/undo_button.png") {
            @Override
            public void clicked() {
                Game.this.undo();
            }
        };
    }

    @Override
    public void setup() {
        //this ensures that the following code only runs when initially entering game
        if (!this.gameLoaded) {
            this.discovered.clear();
            this.history.clear();

            Group.reset();
            Element.reset();

            if (this.saveFile == null) { //new game
                this.mode = "normal";
                ArrayList<Element> elements = new ArrayList<>();
                for (Pack pack : main.packsRoom.getLoadedPacks()) {
                    pack.getStartingElements(elements);
                }
                for (Element element : elements) {
                    this.addElement(element);
                }
                this.hintTime = LocalDateTime.now().plusMinutes(3);
            } else {
                this.mode = this.saveFile.getJson().hasKey("mode") ? this.saveFile.getJson().getString("mode") : "normal";
                JSONArray array = this.saveFile.getJson().getJSONArray("elements");
                for (int i = 0; i < array.size(); i++) {
                    if (!this.addElement(array.getString(i))) {
                        System.err.println("Error: " + array.getString(i) + " could not be loaded from save!");
                        main.loadGame.failed();
                        return;
                    }
                }
                this.hintTime = this.saveFile.getJson().hasKey("hint time") ? LocalDateTime.now().plus(Duration.parse(this.saveFile.getJson().getString("hint time"))) : LocalDateTime.now().plusMinutes(3);
            }

            this.gameLoaded = true;
            this.success.setActive(false);
            this.hint.setActive(false);
        }
    }

    public void exitGame() {
        if (this.saveFile != null && this.saveFile.getJson().getJSONArray("elements").size() < this.getNumberOfElements()) {
            int result = JOptionPane.showConfirmDialog(null, Language.getLanguageSelected().getLocalizedString("game", "not saved"), Language.getLanguageSelected().getLocalizedString("misc", "warning"), JOptionPane.OK_CANCEL_OPTION, JOptionPane.WARNING_MESSAGE);
            if (result == JOptionPane.CANCEL_OPTION) {
                return;
            }
        }
        main.switchRoom(main.menu);
    }

    @Override
    public void draw() {
        main.noStroke();
        main.fill(0);
        main.rect(0, 0, main.screenWidth, main.screenHeight);

        main.textFont(main.font, 20);

        main.fill(255);
        main.textAlign(PConstants.LEFT, PConstants.TOP);
        main.text(Language.getLanguageSelected().getLocalizedString("game", "elements") + ": " + this.getNumberOfElements() + getGap() +
                Language.getLanguageSelected().getLocalizedString("game", "groups") + ": " + this.discovered.keySet().size() + getGap() +
                Language.getLanguageSelected().getLocalizedString("game", "hint timer") + ": " + this.getTimeString(), 10, 10);

        Group.drawGroups();
        Element.drawElements();

        this.groupLeftArrow.draw(Group.GROUP_X, Group.GROUP_Y + (Group.SIZE + Group.GAP) * Group.groupCountY);
        this.groupRightArrow.draw((Group.SIZE + Group.GAP) * Group.groupCountX - Arrow.SIZE, Group.GROUP_Y + (Group.SIZE + Group.GAP) * Group.groupCountY);

        this.elementAUpArrow.draw(main.screenWidth - 20 - Arrow.SIZE, Group.groupSelectedAY);
        this.elementADownArrow.draw(main.screenWidth - 20 - Arrow.SIZE, Group.groupSelectedBY - Group.GAP - Arrow.SIZE);

        this.elementBUpArrow.draw(main.screenWidth - 20 - Arrow.SIZE, Group.groupSelectedBY);
        this.elementBDownArrow.draw(main.screenWidth - 20 - Arrow.SIZE, Group.groupSelectedBY * 2 - Group.GAP - Arrow.SIZE - Group.GROUP_Y);

        if (this.success.isActive()) {
            this.success.draw(main.screenWidth / 2F - this.success.getWidth() / 2F, main.screenHeight / 2F - this.success.getHeight() / 2F);
            Element.drawCreatedElements();
        }

        if (this.hint.isActive()) {
            this.hint.draw(main.screenWidth / 2F - this.hint.getWidth() / 2F, main.screenHeight / 2F - this.hint.getHeight() / 2F);
            Element.drawHintElement(this.hintElement);
        }

        //clear multi-select
        if (!isShiftHeld()) {
            if (Element.getElementsSelected().size() > 0) {
                Element.checkForMultiCombos();
            }
            Element.getElementsSelected().clear();

            int x = main.screenWidth - IconButton.GAP - IconButton.SIZE;
            int y = main.screenHeight - IconButton.GAP - IconButton.SIZE;
            this.exit.draw();
            this.save.draw(x - (IconButton.SIZE + IconButton.GAP), y);
            this.historyButton.draw(x - (IconButton.SIZE + IconButton.GAP) * 2, y);
            this.hintButton.draw(x - (IconButton.SIZE + IconButton.GAP) * 3, y);
            if (this.mode.equals("puzzle")) {
                this.undoButton.draw(x - (IconButton.SIZE + IconButton.GAP) * 4, y);
            }
        }
    }

    public void setSaveFile(SaveFile saveFile) {
        this.saveFile = saveFile;
        this.gameLoaded = false;
    }

    public void saveGame() {
        if (this.saveFile == null) {
            this.saveFile = new SaveFile(main.saveRoom.getSaveName(), new JSONObject());
        }

        JSONArray array = new JSONArray();
        for (ArrayList<Element> list : this.discovered.values()) {
            for (Element element : list) {
                array.append(element.getName());
            }
        }
        this.saveFile.getJson().put("elements", array);

        this.saveFile.getJson().put("last modified", LocalDateTime.now().format(main.formatter));
        this.saveFile.getJson().put("hint time", Duration.between(LocalDateTime.now(), this.hintTime).toString());

        main.saveJSONObject(this.saveFile.getJson(), "resources/saves/" + this.saveFile.getName() + ".json", "indent=4");
    }

    @SuppressWarnings("BooleanMethodIsAlwaysInverted")
    boolean isHintReady() {
        return Duration.between(LocalDateTime.now(), this.hintTime).isZero() || Duration.between(LocalDateTime.now(), this.hintTime).isNegative();
    }

    String getTimeString() {
        //TODO: maybe separate the update time logic?
        Duration duration = Duration.between(LocalDateTime.now(), this.hintTime);
        if (duration.isNegative()) {
            duration = Duration.ZERO;
        }
        int minutes = (int) duration.getSeconds() / 60;
        String minutesString;
        if (minutes < 10) {
            minutesString = "0" + minutes;
        } else {
            minutesString = Integer.toString(minutes);
        }
        int seconds = (int) duration.getSeconds() - minutes * 60;
        String secondsString;
        if (seconds < 10) {
            secondsString = "0" + seconds;
        } else {
            secondsString = Integer.toString(seconds);
        }
        return minutesString + ":" + secondsString;
    }

    private int getNumberOfElements() {
        int sum = 0;
        for (ArrayList<Element> list : this.discovered.values()) {
            sum += list.size();
        }
        return sum;
    }

    private static String getGap() {
        return "          ";
    }

    private boolean addElement(String name) {
        Element element = Element.getElement(name);
        if (element == null) {
            return false;
        }
        this.addElement(element);
        return true;
    }

    public boolean isDiscovered(String element) {
        Group group = main.elements.get(element);
        if (!this.discovered.containsKey(group)) {
            return false;
        }
        for (Element e : this.discovered.get(group)) {
            if (e.getName().equals(element)) {
                return true;
            }
        }
        return false;
    }

    public ArrayList<Combo> getHistory() {
        return this.history;
    }

    public TreeMap<Group, ArrayList<Element>> getDiscovered() {
        return this.discovered;
    }

    private void undo() {
        if (this.history.size() > 0) {
            ArrayList<String> ingredients = this.history.get(this.history.size() - 1).getIngredients();
            ArrayList<String> created = new ArrayList<>();
            //we need to get all combos that had the same ingredients, because the same ingredients can trigger multiple combos
            int i;
            for (i = this.history.size() - 1; i >= 0; i--) {
                Combo combo = this.history.get(i);
                if (CollectionUtils.isEqualCollection(ingredients, combo.getIngredients())) {
                    for (int j = 0; j < combo.getAmount(); j++) {
                        created.add(combo.getElement());
                    }
                } else {
                    break;
                }
            }
            //remove from history
            for (int j = this.history.size() - 1; j > i; j--) {
                this.history.remove(this.history.size() - 1);
            }
            for (String ingredient : ingredients) {
                if (!Objects.requireNonNull(Element.getElement(ingredient)).isPersistent()) {
                    this.addElement(ingredient);
                }
            }
            for (String element : created) {
                this.removeElement(element);
            }
            Element.updateA();
            Element.updateB();
        }
    }

    public void addElement(Element element) {
        Group group = element.getGroup();
        boolean addElement = false;
        if (this.mode.equals("normal")) {
            addElement = !this.discovered.containsKey(group) || !this.isDiscovered(element.getName());
        } else if (this.mode.equals("puzzle")) {
            addElement = !(element.isPersistent() && this.discovered.containsKey(group) && this.isDiscovered(element.getName()));
        }
        if (addElement) {
            if (!this.discovered.containsKey(group)) {
                this.discovered.put(group, new ArrayList<>(Collections.singletonList(element)));
            } else {
                this.discovered.get(group).add(element);
            }
        }
        this.discovered.get(group).sort(Element::compareTo);
    }

    //direct Element objects cannot be used because they are copies
    public void removeElement(String string) {
        Group group = main.elements.get(string);
        //using iterator here because we are only removing one occurrence
        Iterator<Element> iterator = this.discovered.get(group).iterator();
        while (iterator.hasNext()) {
            Element element = iterator.next();
            if (element.getName().equals(string)) {
                iterator.remove();
                break;
            }
        }
        if (this.discovered.get(group).size() == 0) {
            this.discovered.remove(group);
        }
    }

    @Override
    public void mousePressed() {
        this.exit.mousePressed();
        this.save.mousePressed();
        this.historyButton.mousePressed();
        this.hintButton.mousePressed();
        this.undoButton.mousePressed();

        this.groupLeftArrow.mousePressed();
        this.groupRightArrow.mousePressed();

        this.elementAUpArrow.mousePressed();
        this.elementADownArrow.mousePressed();
        this.elementBUpArrow.mousePressed();
        this.elementBDownArrow.mousePressed();

        if (!this.success.isActive() && !this.hint.isActive()) {
            for (Group group : Group.getGroups()) {
                group.mousePressed();
            }

            if (Group.groupSelectedA != null) {
                Group.groupSelectedA.mousePressed();
                for (Element element : Element.getElementsA()) {
                    element.mousePressed();
                }
            }

            if (Group.groupSelectedB != null) {
                Group.groupSelectedB.mousePressed();
                for (Element element : Element.getElementsB()) {
                    element.mousePressed();
                }
            }

            Element.checkForCombos();
        } else {
            this.success.mousePressed();
            this.hint.mousePressed();
        }
    }

    public SaveFile getSaveFile() {
        return this.saveFile;
    }

    private static boolean isShiftHeld() {
        return main.keyPressed && main.keyCode == PConstants.SHIFT;
    }

    @Override
    public void keyPressed() {
        if (main.key == PConstants.CODED) {
            if (main.keyCode == PConstants.LEFT) {
                this.groupLeftArrow.clicked();
            } else if (main.keyCode == PConstants.RIGHT) {
                this.groupRightArrow.clicked();
            } else if (main.keyCode == PConstants.UP) {
                try {
                    main.hintRoom.getElementHint();
                } catch (Hint.NoHintAvailable ignored) {
                }
            }
        } else if (main.key == 'c') {
            for (String element : main.elements.keySet()) {
                this.addElement(element);
            }
        }
    }

    public void success() {
        this.success.setActive(true);
    }

    void setHintElement(Element hintElement) {
        this.hintElement = hintElement;
        this.hint.setActive(true);
    }

    void resetHint() {
        this.hintTime = LocalDateTime.now().plusMinutes(3);
    }

}

File: src/main/rooms/Game.kt
package main.rooms

import main.Element
import main.Language
import main.buttons.*
import main.buttons.ElementButton
import main.buttons.iconbuttons.Exit
import main.buttons.iconbuttons.IconButton
import main.buttons.iconbuttons.Save
import main.combos.Combo
import main.rooms.HintRoom.NoHintAvailable
import main.rooms.PacksRoom.loadedPacks
import org.apache.commons.collections4.CollectionUtils
import org.apache.commons.lang3.time.DurationFormatUtils
import processing.core.PConstants
import processing.data.JSONObject
import java.lang.IllegalArgumentException
import java.time.Duration
import java.time.LocalDateTime
import javax.swing.JOptionPane
import kotlin.collections.ArrayList
import kotlin.math.min

object Game : Room() {

    val discovered = sortedMapOf<Group, ArrayList<Element>>()
    val history = mutableListOf<Combo>()
    var saveFile: SaveFile? = null
    private val success: Pane
    private val hint: Pane
    private val save: Save
    private val exit: Exit
    private val historyButton: IconButton
    private val hintButton: IconButton
    private val undoButton: IconButton
    private val groupLeftArrow: Arrow
    private val groupRightArrow: Arrow
    private val elementAUpArrow: Arrow
    private val elementADownArrow: Arrow
    private val elementBUpArrow: Arrow
    private val elementBDownArrow: Arrow
    private lateinit var hintTime: LocalDateTime //time of next hint
    var gameLoaded = false
    private var hintElement: ElementButton? = null

    @JvmField
    var mode = GameMode.NORMAL
    val isHintReady: Boolean
        get() = Duration.between(LocalDateTime.now(), hintTime).isZero || Duration.between(LocalDateTime.now(), hintTime).isNegative

    //TODO: maybe separate the update time logic?
    fun getTimeString(): String {
        var duration = Duration.between(LocalDateTime.now(), hintTime)
        if (duration.isNegative) {
            duration = Duration.ZERO
        }
        return DurationFormatUtils.formatDuration(duration.toMillis(), "mm:ss")
    }

    private val numberOfElements: Int
         get() {
            var sum = 0
            for (list in discovered.values) {
                sum += list.size
            }
            return sum
        }

    private const val gap: String = "          "
    private val isShiftHeld: Boolean get() = main.keyPressed && main.keyCode == PConstants.SHIFT

    init {
        success = object : Pane() {
            override fun getText(): String {
                return Language.getLanguageSelected().getLocalizedString("game", "you created")
            }
        }
        hint = object : Pane() {
            override fun getText(): String {
                return Language.getLanguageSelected().getLocalizedString("game", "element hint")
            }
        }
        save = Save()
        exit = Exit()
        groupLeftArrow = object : Arrow(LEFT) {
            override fun canDraw(): Boolean {
                return Group.pageNumber > 0
            }

            override fun clicked() {
                if (canDraw()) {
                    Group.pageNumber--
                }
            }
        }
        groupRightArrow = object : Arrow(RIGHT) {
            override fun canDraw(): Boolean {
                return Group.pageNumber < Group.totalPages - 1
            }

            override fun clicked() {
                if (canDraw()) {
                    Group.pageNumber++
                }
            }
        }
        elementAUpArrow = object : Arrow(UP) {
            override fun canDraw(): Boolean {
                return Group.groupSelectedA != null && ElementButton.pageNumberA > 0
            }

            override fun clicked() {
                if (canDraw()) {
                    ElementButton.pageNumberA--
                }
            }
        }
        elementADownArrow = object : Arrow(DOWN) {
            override fun canDraw(): Boolean {
                return Group.groupSelectedA != null && ElementButton.pageNumberA < ElementButton.totalPagesA - 1
            }

            override fun clicked() {
                if (canDraw()) {
                    ElementButton.pageNumberA++
                }
            }
        }
        elementBUpArrow = object : Arrow(UP) {
            override fun canDraw(): Boolean {
                return Group.groupSelectedB != null && ElementButton.pageNumberB > 0
            }

            override fun clicked() {
                if (canDraw()) {
                    ElementButton.pageNumberB--
                }
            }
        }
        elementBDownArrow = object : Arrow(DOWN) {
            override fun canDraw(): Boolean {
                return Group.groupSelectedB != null && ElementButton.pageNumberB < ElementButton.totalPagesB - 1
            }

            override fun clicked() {
                if (canDraw()) {
                    ElementButton.pageNumberB++
                }
            }
        }
        historyButton = object : IconButton("resources/images/history_button.png") {
            override fun clicked() {
                main.switchRoom(HistoryRoom)
            }
        }
        hintButton = object : IconButton("resources/images/hint_button.png") {
            override fun clicked() {
                main.switchRoom(HintRoom)
            }
        }
        undoButton = object : IconButton("resources/images/undo_button.png") {
            override fun clicked() {
                undo()
            }
        }
    }

    override fun setup() {
        //this ensures that the following code only runs when initially entering game
        if (!gameLoaded) {
            discovered.clear()
            history.clear()
            Group.reset()
            Element.reset()
            ElementButton.reset()
            if (saveFile == null) { //new game
                mode = GameMode.NORMAL
                val elements = ArrayList<Element>()
                for (pack in loadedPacks) {
                    pack.getStartingElements(elements)
                }
                for (element in elements) {
                    this.addElement(element)
                }
                hintTime = LocalDateTime.now().plusMinutes(3)
            } else {
                mode = if (saveFile!!.json.hasKey("mode")) {
                    try {
                        GameMode.valueOf(saveFile!!.json.getString("mode"))
                    } catch (e: IllegalArgumentException) {
                        GameMode.NORMAL
                    }
                } else {
                    GameMode.NORMAL
                }
                val array = saveFile!!.json.getJSONArray("elements").stringArray
                for (element in array) {
                    if (!this.addElement(element)) {
                        System.err.println("Error: $element could not be loaded from save!")
                        LoadGame.failed()
                        return
                    }
                }
                hintTime = if (saveFile!!.json.hasKey("hint time")) LocalDateTime.now().plus(Duration.parse(saveFile!!.json.getString("hint time"))) else LocalDateTime.now().plusMinutes(3)
            }
            gameLoaded = true
            success.isActive = false
            hint.isActive = false
        }
    }

    fun exitGame() {
        if (saveFile != null && saveFile!!.json.getJSONArray("elements").size() < numberOfElements) {
            val result = JOptionPane.showConfirmDialog(null, Language.getLanguageSelected().getLocalizedString("game", "not saved"), Language.getLanguageSelected().getLocalizedString("misc", "warning"), JOptionPane.OK_CANCEL_OPTION, JOptionPane.WARNING_MESSAGE)
            if (result == JOptionPane.CANCEL_OPTION) {
                return
            }
        }
        main.switchRoom(Menu)
    }

    override fun draw() {
        main.textFont(main.font, 20f)
        main.fill(255)
        main.textAlign(PConstants.LEFT, PConstants.TOP)
        main.text(Language.getLanguageSelected().getLocalizedString("game", "elements") + ": " + numberOfElements + gap +
                Language.getLanguageSelected().getLocalizedString("game", "groups") + ": " + discovered.keys.size + gap +
                Language.getLanguageSelected().getLocalizedString("game", "hint timer") + ": " + getTimeString(), 10f, 10f)
        Group.drawGroups()
        ElementButton.drawElements()
        groupLeftArrow.draw(Group.GROUP_X.toFloat(), (Group.GROUP_Y + (Group.SIZE + Group.GAP) * Group.groupCountY).toFloat())
        groupRightArrow.draw(((Group.SIZE + Group.GAP) * Group.groupCountX - Arrow.SIZE).toFloat(), (Group.GROUP_Y + (Group.SIZE + Group.GAP) * Group.groupCountY).toFloat())
        elementAUpArrow.draw((main.screenWidth - 20 - Arrow.SIZE).toFloat(), Group.groupSelectedAY.toFloat())
        elementADownArrow.draw((main.screenWidth - 20 - Arrow.SIZE).toFloat(), (Group.groupSelectedBY - Group.GAP - Arrow.SIZE).toFloat())
        elementBUpArrow.draw((main.screenWidth - 20 - Arrow.SIZE).toFloat(), Group.groupSelectedBY.toFloat())
        elementBDownArrow.draw((main.screenWidth - 20 - Arrow.SIZE).toFloat(), (Group.groupSelectedBY * 2 - Group.GAP - Arrow.SIZE - Group.GROUP_Y).toFloat())
        if (success.isActive) {
            success.draw(main.screenWidth / 2f - success.width / 2f, main.screenHeight / 2f - success.height / 2f)
            ElementButton.drawCreatedElements()
        }
        if (hint.isActive) {
            hint.draw(main.screenWidth / 2f - hint.width / 2f, main.screenHeight / 2f - hint.height / 2f)
            ElementButton.drawHintElement(hintElement!!)
        }

        //clear multi-select
        if (!isShiftHeld) {
            if (Element.elementsSelected.isNotEmpty()) {
                Element.checkForMultiCombos()
            }
            ElementButton.elementsSelected.clear()
            Element.elementsSelected.clear()
            val x = main.screenWidth - IconButton.GAP - IconButton.SIZE
            val y = main.screenHeight - IconButton.GAP - IconButton.SIZE
            exit.draw()
            save.draw((x - (IconButton.SIZE + IconButton.GAP)).toFloat(), y.toFloat())
            historyButton.draw((x - (IconButton.SIZE + IconButton.GAP) * 2).toFloat(), y.toFloat())
            hintButton.draw((x - (IconButton.SIZE + IconButton.GAP) * 3).toFloat(), y.toFloat())
            if (mode == GameMode.PUZZLE) {
                undoButton.draw((x - (IconButton.SIZE + IconButton.GAP) * 4).toFloat(), y.toFloat())
            }
        }
    }

    fun saveGame() {
        if (saveFile == null) {
            saveFile = SaveFile(SaveRoom.saveName, JSONObject())
        }
        saveFile!!.json.put("elements", discovered.values.flatten().map { it.id }.toTypedArray())
        saveFile!!.json.put("last modified", LocalDateTime.now().format(main.formatter))
        saveFile!!.json.put("hint time", Duration.between(LocalDateTime.now(), hintTime).toString())
        main.saveJSONObject(saveFile!!.json, "resources/saves/${saveFile!!.name}.json", "indent=4")
    }

    private fun addElement(name: String): Boolean {
        val element: Element = Element.getElement(name) ?: return false
        this.addElement(element)
        return true
    }

    fun isDiscovered(id: String): Boolean {
        val group: Group = main.elements[id]!!
        if (!discovered.containsKey(group)) {
            return false
        }
        for (e in discovered[group]!!) {
            if (e.id == id) {
                return true
            }
        }
        return false
    }

    private fun undo() {
        if (history.isNotEmpty()) {
            val ingredients = history.last().ingredients
            val created = ArrayList<String>()
            //we need to get all combos that had the same ingredients, because the same ingredients can trigger multiple combos
            var i = history.lastIndex
            for ((index, combo) in history.reversed().withIndex()) {
                if (CollectionUtils.isEqualCollection(ingredients, combo.ingredients)) {
                    for (j in 0 until combo.amount) {
                        created.add(combo.element)
                    }
                } else {
                    i = index
                    break
                }
            }
            //remove from history
            for (j in history.lastIndex..i) {
                history.removeLast()
            }
            for (ingredient in ingredients) {
                if (!Element.getElement(ingredient)!!.isPersistent) {
                    this.addElement(ingredient)
                }
            }
            for (element in created) {
                removeElement(element)
            }
            ElementButton.updateA()
            ElementButton.updateB()
        }
    }

    fun addElement(element: Element) {
        val group = element.group
        val addElement = if (mode == GameMode.NORMAL) {
           !discovered.containsKey(group) || !isDiscovered(element.id)
        } else {
            !(element.isPersistent && discovered.containsKey(group) && isDiscovered(element.id))
        }
        if (addElement) {
            if (!discovered.containsKey(group)) {
                discovered[group] = ArrayList()
            }
            discovered[group]!!.add(element)
        }
        discovered[group]!!.sort()
    }

    //direct Element objects cannot be used because they are copies
    fun removeElement(id: String) {
        val group: Group = main.elements[id]!!
        //edge case: if the elements used are the same element, and they are the last remaining elements of the group
        //then when the first element is removed, the group will be removed, so when we call this method for the second element, the group will be removed already
        discovered[group]?.let {
            //using iterator here because we are only removing one occurrence
            val iterator = it.iterator()
            while (iterator.hasNext()) {
                val element = iterator.next()
                if (element.id == id) {
                    iterator.remove()
                    break
                }
            }
            if (it.isEmpty()) {
                discovered.remove(group)
            }
        }
    }

    override fun mousePressed() {
        exit.mousePressed()
        save.mousePressed()
        historyButton.mousePressed()
        hintButton.mousePressed()
        undoButton.mousePressed()
        groupLeftArrow.mousePressed()
        groupRightArrow.mousePressed()
        elementAUpArrow.mousePressed()
        elementADownArrow.mousePressed()
        elementBUpArrow.mousePressed()
        elementBDownArrow.mousePressed()
        if (!success.isActive && !hint.isActive) {
            for (group in Group.getCurrentPageGroups()) {
                group.mousePressed()
            }
            Group.groupSelectedA?.let {
                it.mousePressed()
                for (element in ElementButton.getCurrentPageElements(ElementButton.elementButtonsA, ElementButton.pageNumberA)) {
                    element.mousePressed()
                }
            }
            Group.groupSelectedB?.let {
                it.mousePressed()
                for (element in ElementButton.getCurrentPageElements(ElementButton.elementButtonsB, ElementButton.pageNumberB)) {
                    element.mousePressed()
                }
            }
            ElementButton.checkForCombos()
        } else {
            success.mousePressed()
            hint.mousePressed()
        }
    }

    override fun keyPressed() {
        if (main.key.toInt() == PConstants.CODED) {
            if (main.keyCode == PConstants.LEFT) {
                groupLeftArrow.clicked()
            } else if (main.keyCode == PConstants.RIGHT) {
                groupRightArrow.clicked()
            } else if (main.keyCode == PConstants.UP) {
                try {
                    HintRoom.getElementHint()
                } catch (ignored: NoHintAvailable) {
                }
            }
        } else if (main.key == 'c') {
            for (element in main.elements.keys) {
                this.addElement(element)
            }
        }
    }

    fun success() {
        success.isActive = true
    }

    fun setHintElement(element: Element) {
        this.hintElement = ElementButton(element)
        hint.isActive = true
    }

    fun resetHint() {
        hintTime = LocalDateTime.now().plusMinutes(3)
    }

}

enum class GameMode {
    NORMAL,
    PUZZLE
}